lib/screens/v2/disruption_diagram/builder.ex

```diff
@@ -0,0 +1,967 @@
+        builder
+        |> do_omit(right_omission)
+        |> do_omit(left_omission)
+
+      {:ok, builder}
+    end
+  end
+
+  # Evenly pulls indices from the left and right segments until acc contains enough indices.
+  defp select_split_omission_indices(segment_pair, omit_count, l_acc \\ [], r_acc \\ [])
+
+  defp select_split_omission_indices({l, r}, omit_count, l_acc, r_acc) do
+    select_split_omission_indices(l, r, omit_count, l_acc, r_acc)
+  end
+
+  defp select_split_omission_indices(_l, _r, 0, l_acc, r_acc), do: {l_acc, r_acc}
+
+  defp select_split_omission_indices([], [h | t], n, l_acc, r_acc) do
+    select_split_omission_indices([], t, n - 1, l_acc, [h | r_acc])
+  end
+
+  defp select_split_omission_indices([h | t], [], n, l_acc, r_acc) do
+    select_split_omission_indices(t, [], n - 1, [h | l_acc], r_acc)
+  end
+
+  defp select_split_omission_indices([h | t], r, n, l_acc, r_acc)
+       when length(l_acc) <= length(r_acc) do
+    select_split_omission_indices(t, r, n - 1, [h | l_acc], r_acc)
+  end
+
+  defp select_split_omission_indices(l, [h | t], n, l_acc, r_acc) do
+    select_split_omission_indices(l, t, n - 1, l_acc, [h | r_acc])
+  end
+
+  defp find_safe_segment(original_omission, important_indices, side, offset \\ 1)
+
+  defp find_safe_segment(original_omission, important_indices, :left, offset) do
+    _l..r//1 = original_omission
+
+    tl..tr//1 = tentative_omission = Range.shift(original_omission, -offset)
+
+    if tl <= Vector.first(important_indices) or tr >= Vector.last(important_indices) do
+      :error
+    else
+      first_overlap =
+        important_indices
+        |> Vector.reverse()
+        |> Aja.Enum.find(&(&1 in tentative_omission))
+
+      case first_overlap do
+        nil ->
+          {:ok, tentative_omission, offset}
+
+        i ->
+          # The tentative window contains an important index. Move the window past the first important index and try again.
+          find_safe_segment(original_omission, important_indices, :left, 1 + r - i)
+      end
+    end
+  end
+
+  defp find_safe_segment(original_omission, important_indices, :right, offset) do
+    l.._r//1 = original_omission
+
+    tl..tr//1 = tentative_omission = Range.shift(original_omission, offset)
+
+    if tl <= Vector.first(important_indices) or tr >= Vector.last(important_indices) do
+      :error
+    else
+      first_overlap = Aja.Enum.find(important_indices, &(&1 in tentative_omission))
+
+      case first_overlap do
+        nil ->
+          {:ok, tentative_omission, offset}
+
+        i ->
+          # The tentative window contains an important index. Move the window past the first important index and try again.
+          find_safe_segment(original_omission, important_indices, :right, 1 + i - l)
+      end
+    end
+  end
+
+  defp do_add_slots(builder, 0, _), do: builder
+
+  defp do_add_slots(builder, _greater_than_0, _)
+       when vec_size(builder.left_end) == 0 and vec_size(builder.right_end) == 0 do
```

Comment:
OH! I think I understand what left_end and right_end are for -- they have all the stops not currently going into the rest of the diagram? If that's the case, can that be a comment in the top of this file where left_end / right_end are defined?